concepts/algorithms/covering_edges.py

```diff
@@ -0,0 +1,84 @@
+"""Covering Edges
+
+cf. Carpineto, Claudio, and Giovanni Romano.
+Concept data analysis: Theory and applications.
+John Wiley & Sons, 2004.
+"""
+
+
+def covering_edges(concept_list, context):
+    """Yield mapping edge as ``((extent, intent), (lower_extent, lower_intent))``
+    pairs (concept and it's lower neighbor) from ``context`` and ``concept_list``
+    
+    Example:
+        >>> from concepts import make_context, ConceptList
+        >>> from concepts._common import Concept
+
+        >>> context = make_context('''
+        ...  |0|1|2|3|4|5|
+        ... A|X|X|X| | | |
+        ... B|X| |X|X|X|X|
+        ... C|X|X| | |X| |
+        ... D| |X|X| | | |''')
+
+        >>> concepts = [('ABCD', ''),
+        ...             ('ABC', '0'),
+        ...             ('AC', '01'),
+        ...             ('A', '012'),
+        ...             ('', '012345'),
+        ...             ('C', '014'),
+        ...             ('AB', '02'),
+        ...             ('B', '02345'),
+        ...             ('BC', '04'),
+        ...             ('ACD', '1'),
+        ...             ('AD', '12'),
+        ...             ('ABD', '2')]
+
+        >>> concept_list = ConceptList.frompairs(
+        ...     map(lambda c: (context._Objects.frommembers(c[0]),
```

Comment:
use of private API

Reply:
Any tips how to avoid that? Problem is that this whole function is probably also for internal use only. Maybe the whole example can be defined other way, originally wanted to avoid using `fast_generate_from` in the example, but what do you think? That would remove this line. Example could be context -> `fast_generate_from` -> `ConceptList` -> `covering_edges`.

Reply:
I am confused: Can't the function just use the output of a concept generator like `fcbo` as input?

> context -> fast_generate_from -> ConceptList -> covering_edges.

See the other discussion on avoiding materialization. :)

Although, I might be missing something and need to read the paper first.

> Problem is that this whole function is probably also for internal use only.

In case a test fiddles with internals or otherwise has a complex setup, state requirement, or assertions, it should probably be done in `tests/`. Doctests are better for simple (pure) functions and if they are used as an example, they should only use official API (in general, I think a good API is one that just flows naturally in the repl). As this is a pure function, my hope is we can have the doctest both guard the implementation and serve to understand the API.

Reply:
Another option might be to create the input from scratch in the test (creating the two `bitsets.bitset`s).

Reply:
The reason why context is needed are the prime operators and bitset classes.

Reply:
Okay, I see now, we can supply the `bitsets` for the doctest but they would lack the `.prime()`-methods. So in the doctest, probably use the output from an intent generator. However, not sure why we need the context as an argument if the input is an iterable of `(extent, intent)` pairs (i.e. bitsets with `.prime()`-methods)

Reply:
The algorithm still have to calculate some prime operations, the speed up is based on ability to access existing concepts by extent from dict (in comparation to upper neighbor). So it is not like, that this alg is able to build lattice just from concepts (at least I am not able to see that in pseudocode).

Reply:
> The algorithm still have to calculate some prime operations.

What I was trying to say is that we don't need the `context` for that (AFAICT the current version uses the methods on the given pairs for that). We could pass `Objects` and `Properties` explicitly, but maybe it's enough to just use the  `__class__`es from the first item instead.

Reply:
To illustrate from another domain:

```python
import csv
import itertools

def iterrows():
    yield 'SPAM,EGGS\n'
    yield 'span,EGGS\n'
    
reader = csv.reader(iterrows())

try:
    header = next(reader)
except StopIteration:
    raise ValueError('missing header')

assert header == ['SPAM', 'EGGS']

assert list(reader) == [['span', 'EGGS']]
```

Reply:
Ah, right, that is a clever solution, thanks. I will add that and remove context.

Reply:
I have checked the comments once again. So, do you agree, that `covering_edges` would yield `NamedTuple`'s:

```python
class ConceptMappingInfo(typing.NamedTuple):

    concept: concepts._common.Concept

    lower_neighbors: typing.Tuple[concepts._common.Concept, ...]
```

I am still confused what is the purpose of `concepts._common.Concept` but it would be useful here.

Other option is (hope I got typing for bitset right):

```python
class ConceptMappingInfo(typing.NamedTuple):

    extent: bitsets.bases.BitSet

    intent: bitsets.bases.BitSet

    lower_neighbors: typing.Tuple[typing.Tuple[bitsets.bases.BitSet, bitsets.bases.BitSet], ...]
```

As for input for `covering_edges`, I would be for:

```python
def covering_edges(concept_list: typing.Iterator[typing.Tuple[bitsets.bases.BitSet,
                                                              bitsets.bases.BitSet]]):
    pass
```

In that case, `fast_generate_from` can be directly plugged in. Inside your example (https://github.com/xflr6/concepts/pull/15#discussion_r635093902) would be used.

Reply:
The idea mentioned in https://github.com/xflr6/concepts/pull/15#issuecomment-843531411 is interesting, but I am worried it would be better to build full lattice from `ConceptMappingInfo` in next step.

Reply:
Quick notes:

Still nesting (yield quadruples, see earlier discussion)?
```python
class ConceptMappingInfo(typing.NamedTuple):

    concept: concepts._common.Concept
````
IIUC this differs from `Yield mapping edge as ``((extent, intent), (lower_extent, lower_intent))`, is it intended to include both the extent and the intent of neighbors even though one of both suffices for a reference?

```python
lower_neighbors: typing.Tuple[concepts._common.Concept, ...]
````

Same here:

```python
 lower_neighbors: typing.Tuple[typing.Tuple[bitsets.bases.BitSet, bitsets.bases.BitSet], ...]
```

nit: `typing.Iterable` (prefer generic types for arguments e.g. `Sequence` and concrete ones for returns, e.g. `List`, cf. Postel's law)

```
def covering_edges(extent_intent_pairs: typing.Iterator[typing.Tuple[bitsets.bases.BitSet,
                                                              bitsets.bases.BitSet]]):
```

We might need to do some more to express the requirement that the bitset pairs need to be ones that where for prime in `matrices.py`.

Reply:
You right, that both extent/intent are not required for storing the relationship, so probably something like:

```python
class ConceptMappingInfo(typing.NamedTuple):

    extent: bitsets.bases.BitSet

    intent: bitsets.bases.BitSet

    lower_neighbors_extents: typing.Tuple[bitsets.bases.BitSet, ...]
```

Agree with `typing.Iterable` as input. Maybe we will get better big picture when we will try to build whole lattice with this function? Then we can modify the API more and keep it public or make it fully internal only function.

Do you agree that adding `lattice` function into `covering_edges.py` in the same fashion as in `lindig.py` would be good approach? That function would accept `context` as argument, because internally uses `fast_generate_from` to generate concepts and `covering_edges` to build mapping (that is main difference to the one in the `lindig.py` output will be same). This `lattice` function should be easily pluggable into existing data structure.

Then we can finish the whole code.